File: G_Projects/F_Elevator_System/ObserverPattern/ElevatorObserver.java
package G_Projects.F_Elevator_System.ObserverPattern;

import G_Projects.F_Elevator_System.CommonEnums.ElevatorState;
import G_Projects.F_Elevator_System.UtilityClasses.Elevator;

public interface ElevatorObserver {
    void onElevatorStateChange(Elevator elevator, ElevatorState state); //But elevator already has state info, so do we need it? Refer to LawOfDemeter.md
    void onElevatorFloorChange(Elevator elevator, int floor);
}

File: Ruby/A_Project/C_Elevator_System/ObserverPattern/elevator_observer.rb
# Abstract observer interface (Observer Pattern)
class ElevatorObserver
  def on_elevator_state_change(elevator, state)
    raise NotImplementedError, "#{self.class}#on_elevator_state_change must be implemented"
  end

  def on_elevator_floor_change(elevator, floor)
    raise NotImplementedError, "#{self.class}#on_elevator_floor_change must be implemented"
  end
end
